sticht/rollbacks/base.py

```diff
@@ -0,0 +1,235 @@
+            return ''
+
+    def start_slo_watcher_threads(self, service: str, soa_dir: str) -> None:
+        _, self.slo_watchers = watch_slos_for_service(
+            service=service,
+            individual_slo_callback=self.individual_slo_callback,
+            all_slos_callback=self.all_slos_callback,
+            sfx_api_token=self.get_signalfx_api_token(),
+            soa_dir=soa_dir,
+        )
+
+    def start_metric_watcher_threads(self, service: str, soa_dir: str) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_signalfx_api_token(self) -> str:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_splunk_api_token(self) -> Tuple[str, str]:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def auto_rollbacks_enabled(self) -> bool:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_auto_rollback_delay(self) -> float:
+        raise NotImplementedError()
+
+    def any_slo_failing(self) -> bool:
+        return self.auto_rollbacks_enabled() and self.slo_watchers is not None and any(
+            w.failing for w in self.slo_watchers
+        )
+
+    def any_metric_failing(self) -> bool:
+        return self.auto_rollbacks_enabled() and self.metric_watchers is not None and any(
+            w.failing for w in self.metric_watchers
+        )
+
+    def any_rollback_condition_failing(self) -> bool:
+        return self.any_slo_failing() or self.any_metric_failing()
+
+    def individual_slo_callback(self, label: str, bad: Optional[bool]) -> None:
+        if bad:
+            self.update_slack_thread(f'SLO started failing: {label}', color='danger')
+        else:
+            self.update_slack_thread(f'SLO is now OK: {label}', color='good')
+
+    def all_slos_callback(self, bad: bool) -> None:
+        if bad:
+            self.trigger('slos_started_failing')
+        else:
+            self.trigger('slos_stopped_failing')
+        self.update_slack()
+
+    def individual_metric_callback(self, label: str, bad: Optional[bool]) -> None:
+        if bad:
+            self.update_slack_thread(f'metric started failing: {label}', color='danger')
+        else:
+            self.update_slack_thread(f'metric is now OK: {label}', color='good')
+
+    def all_metrics_callback(self, bad: bool) -> None:
+        if bad:
+            self.trigger('metrics_started_failing')
+        else:
+            self.trigger('metrics_stopped_failing')
+        self.update_slack()
+
+    # TODO: figure out what to do wrt to these triggers now that we have SLOs and metric query
+    # rollacks
+
+    def start_auto_rollback_countdown(self, extra_text) -> None:
+        self.start_timer(
+            self.get_auto_rollback_delay(),
+            'rollback_slo_failure',
+            'automatically roll back',
+            extra_text=extra_text,
+        )
+
+    def cancel_auto_rollback_countdown(self) -> None:
+        self.cancel_timer('rollback_slo_failure')
```

Comment:
not sure what we want to do here yet with these timers so i've left them alone for now